wg-api/spec-documents/better-native-window-open.md

```diff
@@ -0,0 +1,105 @@
+## Background
+### Impetus
+
+In the current implementation, we destroy the newly-created `WebContents` when the consumer calls `event.preventDefault()`. In recent Electron versions, this can cause a crash, because the WebContents is still initializing when we kill it. Additionally, the preload script has started to run, which has the potential to cause quite a bit of confusion for developers looking at logs. We could patch and continue to patch these weirdnesses, but fixing the root seems appropriate sooner rather than later.
+
+### Future of window.open
+
+As we align more with Chrome's process model, `BrowserWindowProxy` will become less necessary. `nativeWindowOpen: true` will become the default and the ergonomics (and durability) of this API will matter quite a bit more!
+
+### Current API pain points
+
+1. Returning a window via `event.newGuest` in the event handler is unprecedented in Electron and doesn't fit normal "Web Events" semantics
+2. The `options` argument of `new-window` is a mix of parsed options from the feature string and important fields provided by Electron, like `webContents` itself. It's not clear that you need to pass `webContents` into the overridden window for things to work.
+3. Can't set `webPreferences`, since the webContents was already created
+
+### Use cases
+
+1. Custom handling for target="_blank" links from trusted or untrusted content.
+    - Canceling disallowed urls
+    - Kicking out to a browser vs. opening windows with Electron
+2. Opening application windows with synchronous DOM access
+3. Navigating to external sites (for OAuth flows, for example)
+
+### Platforms
+
+All platforms
+
+## Proposal
+
+### Summary
+
+At a high level, I propose we add an event, `will-create-child-web-contents`, so that we can cancel the renderer-initialized `WebContents` creation earlier than we do now. Additionally, for improved clarity and ergonomics, I suggest we change the name of the `WebContents`'s  `new-window` event to `will-create-child-window` and slightly change its mechanics. 
+
+I would guess that most classes of app would need to subscribe to zero or one of these events. Only the most complex would need both (Slack! Teams!).
+
+### API Design
+
+#### `will-create-child-web-contents`
+
+Addresses parts of **use case #1** wherein the consumer wants to stop a renderer's `window.open` call from succeeding.
+
+**Note:** this will not fire for an iframe, a potential point of confusion.
+
+**Possible implementation**
+
+There are already delegate methods that we could hook into to emit this event before the `WebContents` is created, no patching necessary: `ShouldCreateWebContents` (7-0-x) and `IsWebContentsCreationOverridden` (master). 
+
+**Example**
+
+```javascript
+mainWindow.webContents.on(
+  'will-create-child-web-contents',
+  (event, url, frameName, disposition, options, additionalFeatures) =>
```

Comment:
It seems like it would be useful to know what WebContents the new WebContents will child of.

Reply:
I believe it will always be `mainWindow.webContents`, i.e. the one you're listening to?

Reply:
ah, okay then!